test/e2e/storage/testsuites/snapshottable_metrics.go

```diff
@@ -0,0 +1,406 @@
+		if actualValue < expectedValue {
+			framework.Logf("metric value for key %s invalid: expected at least %v, got %v", metricKey, expectedValue, actualValue)
+			return false
+		}
+
+		return true
+	}); successful {
+		return
+	}
+
+	framework.Failf("Unable to get snapshot controller metrics after %v", timeout)
+}
+
+func (scm *snapshotControllerMetrics) getSnapshotControllerMetricValue() (float64, error) {
+	metricKey := scm.getMetricKey()
+
+	// grab and parse into readable format
+	err := scm.grabSnapshotControllerMetrics()
+	if err != nil {
+		return 0, err
+	}
+
+	var metrics map[string]float64
+	switch scm.cfg.metricType {
+	case "count":
+		metrics = scm.countMetrics
+
+	case "sum":
+		metrics = scm.sumMetrics
+
+	case "bucket":
+		metrics = scm.bucketMetrics
+
+	default:
+		return 0, fmt.Errorf("invalid metric type %s", scm.cfg.metricType)
+	}
+
+	actual, ok := metrics[metricKey]
+	if !ok {
+		framework.Logf("failed to find metric key %s inside of metrics:", metricKey)
+		for k, v := range metrics {
+			framework.Logf("%s: %v", k, v)
+		}
+		return 0, fmt.Errorf("did not find metric for key %s", metricKey)
+	}
+
+	return actual, nil
+}
+
+func (scm *snapshotControllerMetrics) grabSnapshotControllerMetrics() error {
+	// pull all metrics
+	metrics, err := scm.metricsGrabber.GrabFromSnapshotController()
+	if err != nil {
+		e2eskipper.Skipf("Could not get snapshot controller metrics - skipping")
+		return err
+	}
+
+	for method, samples := range metrics {
+		switch method {
+		case "snapshot_controller_operation_total_seconds_count":
+			for _, sample := range samples {
+				count := float64(sample.Value)
+				scm.countMetrics[snapshotMetricSampleKey(sample, method)] = count
+			}
+
+		case "snapshot_controller_operation_total_seconds_sum":
+			for _, sample := range samples {
+				sum := float64(sample.Value)
+				scm.sumMetrics[snapshotMetricSampleKey(sample, method)] = sum
+			}
+
+		case "snapshot_controller_operation_total_seconds_bucket":
+			for _, sample := range samples {
+				bucketCount := float64(sample.Value)
+				scm.bucketMetrics[snapshotMetricSampleKey(sample, method)] = bucketCount
+			}
+		}
+	}
+
+	return nil
+}
+
+func (scm *snapshotControllerMetrics) getMetricKey() string {
+	return snapshotMetricKey(scm.cfg.metricName, scm.cfg.driverName, scm.cfg.operationName, scm.cfg.operationStatus, scm.cfg.snapshotType, scm.cfg.le)
+}
+
+func snapshotMetricSampleKey(sample *model.Sample, metricName string) string {
+	operationName := string(sample.Metric["operation_name"])
+	driverName := string(sample.Metric["driver_name"])
+	operationStatus := string(sample.Metric["operation_status"])
+	snapshotType := string(sample.Metric["snapshot_type"])
+	le := string(sample.Metric["le"])
+
+	return snapshotMetricKey(metricName, driverName, operationName, operationStatus, snapshotType, le)
+}
+
+func snapshotMetricKey(metricName, driverName, operationName, operationStatus, snapshotType, le string) string {
```

Comment:
how about just:
return fmt.Sprintf("%s_%s_%s_%s" ....)? 
can any of the passed-in strings be empty?

Reply:
Yeah, `le` can be empty. This metric value is only present in the "bucket" metrics for a histogram. I had to construct the key dynamically to get it right.